src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java

```diff
@@ -988,6 +1119,78 @@ private int implDoFinal(ByteBuffer outBuffer)
         }
     }
 
+    private int getBytesToCompleteBlock(int availableBytes) {
+        int partBlock = availableBytes & (blockSize - 1);
+        return partBlock == 0 ? 0 : blockSize - partBlock;
+    }
+
+    private int getCTSMustBeBuffered(int availableBytes) {
+        return Math.min(availableBytes,
+                padBuffer.length - getBytesToCompleteBlock(availableBytes));
+    }
+
+    /**
+     * The ciphertext ordering for the three variants can be depicted as
+     * follows, where 'p' is the penultimate block (which may be partial
+     * or full), and 'f' the full final block:
+     *
+     *                    'p' is a partial block   'p' is a full block
+     *                   ------------------------ ---------------------
+     *   CS1 (NIST)     |     .... pp ffff       |    .... pppp ffff
+     *   CS2 (Schneier) |     .... ffff pp       |    .... pppp ffff
+     *   CS3 (Kerberos) |     .... ffff pp       |    .... ffff pppp
+     *
+     * After encryption, we get the ciphertext from the token formatted as
+     * specified in the SunPKCS11 'cipherTextStealingVariant' configuration
+     * property. Conversely, before decryption, the ciphertext has to be passed
+     * to the token according to the previous formatting. This method converts
+     * the ciphertext between the format used by the token and the one used by
+     * SunJCE's "AES/CTS/NoPadding" implementation (CS3 as described by RFC
+     * 2040, section 8).
+     */
+    private void convertCTSVariant(ByteBuffer ciphertextBuf,
+            byte[] ciphertextArr, int ciphertextEnd) {
+        if (padBufferLen == blockSize) {
+            // No reordering needed for a single block
+            return;
+        }
+        assert token.ctsVariant != null : "CTS algorithms should not be " +
+                "registered if the CTS variant of the token is unknown";
+        if (token.ctsVariant == Token.CTSVariant.CS3) {
+            // Already CS3
+            return;
+        }
+        int pad = padBufferLen % blockSize;
+        if (token.ctsVariant == Token.CTSVariant.CS2 && pad != 0) {
+            // CS2 and 'p' is a partial block, equal to CS3
+            return;
+        }
+        if (ciphertextArr != null) {
+            ciphertextBuf = ByteBuffer.wrap(ciphertextArr);
```

Comment:
Can we add a comment here to caution that position may be incorrect (since the offset is not passed to this call) and thus need to always supply an index for reading/writing values to the `ciphertextBuf`?

Reply:
It's not so much about position being incorrect —`convertCTSVariant` makes no assumptions about it, but could have reset position to the beginning of the last 2 blocks— but that `ciphertextBuf` bytes should not be modified except for the last 2 blocks. I'm okay with adding a comment but  I don't see any extension of the `convertCTSVariant` function that could be at risk of having to access anything other than the last 2 blocks.

Reply:
True, the proposed change does not use the position of the `ciphertextBuf`. 
I'd just like some comment covering this since I find it strange that you call `ByteBuffer.wrap(cipherTextArr)` without the offset when the javadoc states that the position is set to 0. Maybe others may also wonder why. That's all.